### ISSUE TITLE
Hot reload not working with Next.js 13.4.9 in Custom Server development environment

### ISSUE THREAD

**Issue by knightrun** [vercel/next.js#52931]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.16.1
      npm: 9.5.1
      Yarn: 1.22.18
      pnpm: 8.4.0
    Relevant Packages:
      next: 13.4.9
      eslint-config-next: 12.3.4
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.7.4
    Next.js Config:
      output: standalone
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

.

### To Reproduce

1. Clone the project using Next.js 13.4.9 with the app directory structure.
2. Start the Custom Server development environment using the command pnpm dev
4. Access the application in a web browser.
5. Make changes to the application code within the app directory, e.g. update a component or a page.
6. Observe that the changes are not automatically reflected in the browser, and the hot reload does not occur.

### Describe the Bug

I am using Next.js version 13.4.9 with the app directory structure and running my development environment with a Custom Server. However, I've noticed that the hot reload feature is not functioning as expected. Changes made to the code are not automatically reflected in the browser during development.

### Expected Behavior

1. When I make changes to my Next.js application code (e.g., components, pages, styles) in the app directory while the Custom Server is running, I should see the changes applied automatically without needing to manually refresh the browser.
2. The development environment should detect file modifications and trigger a hot reload, ensuring a smooth and efficient development experience.

### Which browser are you using? (if relevant)

Chrome 114.0.5735.198

**Comment by leeyc924** [comment]
There is the same issue when updated to version 13.4.11

**Comment by kacper-zylka** [comment]
Having the same issue. When I make a change in a file the terminal prints 
```
- wait compiling...
- event compiled client and server successfully in 403 ms (2429 modules)
```
but the change is not reflected in the browser until I refresh manually.

I see this error which could be related:
<img width="582" alt="image" src="https://github.com/vercel/next.js/assets/5412047/9a5c4939-6a8f-4081-bb69-229877f2b6db">

**Comment by leeyc924** [comment]
It seems that I encountered an issue while modifying "revalidate" during the upgrade to version **13.3.xx**. When I tested it with version **13.2.4,** it worked correctly.

#https://github.com/vercel/next.js/compare/v13.2.4-canary.4...v13.2.4-canary.5#diff-4c54e369ddb9a2db1eed95fe1d678f94c8e82c540204475d42c78e49bf4f223a

**Comment by Inzendis** [comment]
So im having this issue right now. Any news?

**Comment by Inzendis** [comment]
> Having the same issue. When I make a change in a file the terminal prints
> 
> ```
> - wait compiling...
> - event compiled client and server successfully in 403 ms (2429 modules)
> ```
> 
> but the change is not reflected in the browser until I refresh manually.
> 
> I see this error which could be related: <img alt="image" width="582" src="https://user-images.githubusercontent.com/5412047/256216214-9a5c4939-6a8f-4081-bb69-229877f2b6db.png">

So im having this issue right now. Any news?

I tried using nodemon as well and it is detecting changes but I still can't see the changes even if it refreshes.

**Comment by nwankwo-ikemefuna** [comment]
Upgrading my node version from 16.x to >= 20 fixed it for me. I didn't try with v18.16.1 though.

**Comment by sheisacreative** [comment]
Thanks, @nwankwo-ikemefuna!
It solved for me too. :)

I just installed [node version 20.9.0](https://nodejs.org/en) and it worked like a charm.

**Comment by BorPawel** [comment]
Yeah, same issue, I have node v20.10.0 and next 14.0.4. In all cases which I tested I had pages folder with _app.js file. When I deleted it, it did not help, however, I started new project, and installed and configured all nessesery things but without _app.js file it got fixed. Got these errors in dev tools, idk if it helps.
![image](https://github.com/vercel/next.js/assets/88384089/018e06a7-2a23-4727-a672-29528f98933f)

**Comment by frankievx** [comment]
Upgrading node to v20.10.0 didn't help me. It seems as if my browser is defaulting to my prod version after running a local build than the dev server that's running.

**Comment by Sumansourabh14** [comment]
The following worked for me.

I had to create a `nodemon.json` file in my root directory with the following content:

```
{
  "verbose": true,
  "ignore": ["node_modules", ".next"],
  "watch": ["server/**/*", "index.js"],
  "ext": "js json"
}
```

After this, my Next.js app started reloading after any change made in the app.

Version: 

`"next": "14.1.0"`
`"node": "18.17.0"`

**Comment by VenkateshSavai** [comment]
Title: HMR WebSocket Connection Fails After Upgrading from Next.js 13 to 14

Description: I've recently upgraded my project from Next.js version 13 to 14 and started experiencing WebSocket connection failures for Hot Module Replacement (HMR). The console logs multiple errors indicating that the WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed.

Expected Behavior: HMR should establish a WebSocket connection successfully, allowing for live reloading of modules during development.

Actual Behavior: The browser console logs WebSocket connection errors repeatedly, stating: "WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed".

Environment Information:

Next.js version: 13 (before upgrade), 14 (after upgrade)
Node.js version: 20.10.0
OS: windows
No custom Webpack configuration was used prior to the upgrade.

Attempted Solutions:

Checked next.config.js for any HMR or WebSocket related configurations that might need updating.
Verified that there are no network issues or blockers such as browser extensions or firewall settings.
Searched through the Next.js GitHub issues and discussions for similar problems post-upgrade.
Additional Context:

The issue occurs consistently after the upgrade and was not present in the previous version of Next.js.
Attaching a screenshot of the console errors for reference.
The development server is running on http://localhost:3000 as usual, with no other apparent issues besides HMR.
Screenshot:
![image](https://github.com/vercel/next.js/assets/139204811/745cf279-6c83-4839-af81-24eead61f272)
